odo/backends/url.py

```diff
@@ -0,0 +1,178 @@
+from __future__ import print_function, division, absolute_import
+
+import os
+import uuid
+
+from contextlib import contextmanager, closing
+
+try:
+    from urllib2 import urlopen
+except ImportError:
+    from urllib.request import urlopen
+
+from toolz import memoize
+from toolz.curried import take, map, pipe
+
+from .. import discover
+from ..resource import resource
+from ..append import append
+from ..convert import convert
+from ..temp import Temp, _Temp
+from .csv import CSV
+from .json import JSON, JSONLines
+from .text import TextFile
+
+
+from multipledispatch import MDNotImplementedError
+
+from .text import TextFile
+
+from ..compatibility import urlparse
+from ..utils import tmpfile, ext, sample, filter_kwargs
+
+
+class _URL(object):
+    """ Parent class for data accessed through ``URLs``
+
+    Parameters
+    ----------
+
+    url : str
+        full url to file
+    chunk_size : int (default 1024)
+        Size of chunks streamed into memory
+    decode_unicode : bool (default False)
+        If True, content will be decoded using the best available encoding based on the response.
+
+     Examples
+    --------
+
+    >>> from odo import URL, CSV
+    >>> u = URL(CSV)('http://foo.com/myfile.csv')
+
+    Normally create through resource uris
+
+    >>> data = resource('http://foo.com/myfile.csv')
+    >>> data.url
+    'http://foo.com/myfile.csv'
+    >>> data.filename
+    'myfile.csv'
+    """
+    def __init__(self, url, chunk_size=1024, decode_unicode=False, *args,
+                 **kwargs):
+
+        self.subtype.__init__(self, url, *args, **kwargs)
+
+        self.url = url
+        self.chunk_size = chunk_size
+        self.decode_unicode = decode_unicode
+        self.filename = os.path.basename(urlparse(url).path)
+
+
+def URL(cls):
+    return type('URL(%s)' % cls.__name__, (_URL, cls), {'subtype': cls})
+
+URL.__doc__ = _URL.__doc__
+URL = memoize(URL)
+
+@sample.register((URL(CSV), URL(JSONLines)))
+@contextmanager
+def sample_url_line_delimited(data, lines=5):
+    """Get a size `length` sample from an URL CSV or URL line-delimited JSON.
+
+    Parameters
+    ----------
+    data : URL(CSV)
+        A hosted CSV
+    lines : int, optional, default ``8192``
+        Number of bytes to read into memory
+    """
+
+    with closing(urlopen(data.url)) as r:
+        raw = pipe(r, map(bytes.decode), take(10), list, '\n'.join)
```

Comment:
You don't need to instantiate the sequence with `list` here. `'\n'.join` works on any sequence of strings.